lib/domain.js

```diff
@@ -0,0 +1,214 @@
+                    }
+                }
+
+                if (i === domains.length - 1) {
+                    Domain.create(objects, (err1, doc) => {
+                        if (err1)
+                            reject(err1);
+                        else {
+                            fulfill(objects.map(d => d.id).concat(exists).sort());
+                        }
+                    });
+                }
+            });
+        }
+    });
+}
+
+module.exports.maybeDelete = function (domains) {
+    for (let domain of domains) {
+        Journal.findOne({domains: domain.id}, (err, journal) => {
+            if (!journal) {
+                Domain.findByIdAndDelete(domain.id);
+            }
+        });
+    }
+}
+
+module.exports.getDomainsForPublisher = function (publisher) {
+    return new Promise((fulfill, reject) => {
+        getDomains(publisher)
+            .then(res => {
+                fulfill(res);
+            })
+            .catch(err => {
+                reject(err);
+            });
+    });
+}
+
+module.exports.getDomainsForJournal = function (journal) {
+    return new Promise((fulfill, reject) => {
+        getDomains(journal)
+            .then(res => {
+                fulfill(res);
+            })
+            .catch(err => {
+                reject(err);
+            });
+    });
+}
+
+module.exports.checkExistence = function (domain) {
+    return new Promise((fulfill, reject) => {
+        Domain.findOne({
+            topLevelDomain: domain.topLevelDomain,
+            secondLevelDomain: domain.secondLevelDomain,
+            subDomain: domain.subDomain
+        }, (err, resDomain) => {
+            if (err || !resDomain)
+                reject(domain);
+            else {
+                fulfill(resDomain);
+            }
+        })
+    });
+}
+
+getDomains = function (object) {
+    return new Promise((fulfill, reject) => {
+        let promises = [];
+        let domains = [];
+        for (let domain of object.domains) {
+            promises.push(new Promise((fulfill2, reject2) => {
+                Domain.findOne({id: domain}, '-_id id topLevelDomain secondLevelDomain subDomain', (err, domain) => {
+                    if (err) {
+                        reject2("Error getting domain from database")
+                        return;
+                    }
+                    if (!domain) {
+                        reject2("No domain with this ID")
+                        return;
+                    }
+                    domains.push(domain);
+                    fulfill2();
+                });
+            }));
+        }
+
+        Promise.all(promises)
+            .then(() => {
+                fulfill(domains);
+            })
+            .catch((error) => {
+                reject(error);
+            });
+    });
+}
+
+removeProtocol = async function (domain) {
+    if (domain.includes("://")) {
+        let index = domain.indexOf('://');
+        return domain.substring(index + 3);
```

Comment:
String subsetting is always dangerous! I suggest to use `new URL(..)` and pass around objects, not strings.